File: 01-Programming-Languages/R/OOP/classes.R
# Shape Hierarchy in R
#
# Demonstrates R's different OOP systems:
# - S3 classes (informal, most common)
# - R6 classes (reference-based, Python/Java-like)
#
# Run: Rscript classes.R

# ============================================================
# S3 Classes: Informal OOP (most common in R)
# ============================================================

# Constructor
Shape <- function(name, color = "black") {
  structure(
    list(name = name, color = color),
    class = "Shape"
  )
}

# Methods
print.Shape <- function(x, ...) {
  cat(x$name, "[", x$color, "]\n")
}

summary.Shape <- function(object, ...) {
  cat(object$name, ": area =", round(object$area, 2),
      ", perimeter =", round(object$perimeter, 2), "\n")
}

# Circle S3 class
Circle <- function(radius, color = "black") {
  shape <- Shape("Circle", color)
  shape$radius <- radius
  shape$area <- pi * radius^2
  shape$perimeter <- 2 * pi * radius
  class(shape) <- c("Circle", "Shape")
  shape
}

print.Circle <- function(x, ...) {
  cat("Circle (r =", x$radius, ") [", x$color, "]",
      ": area =", round(x$area, 2),
      ", perimeter =", round(x$perimeter, 2), "\n")
}

# Rectangle S3 class
Rectangle <- function(width, height, color = "black") {
  shape <- Shape("Rectangle", color)
  shape$width <- width
  shape$height <- height
  shape$area <- width * height
  shape$perimeter <- 2 * (width + height)
  class(shape) <- c("Rectangle", "Shape")
  shape
}

print.Rectangle <- function(x, ...) {
  cat("Rectangle (", x$width, "x", x$height, ") [", x$color, "]",
      ": area =", round(x$area, 2),
      ", perimeter =", round(x$perimeter, 2), "\n")
}

# Triangle S3 class
Triangle <- function(a, b, c, color = "black") {
  shape <- Shape("Triangle", color)
  shape$a <- a
  shape$b <- b
  shape$c <- c
  s <- (a + b + c) / 2
  shape$area <- sqrt(s * (s - a) * (s - b) * (s - c))
  shape$perimeter <- a + b + c
  class(shape) <- c("Triangle", "Shape")
  shape
}

print.Triangle <- function(x, ...) {
  cat("Triangle (", x$a, ",", x$b, ",", x$c, ") [", x$color, "]",
      ": area =", round(x$area, 2),
      ", perimeter =", round(x$perimeter, 2), "\n")
}

# S3 generic functions
area <- function(shape) UseMethod("area")
area.Shape <- function(shape) shape$area

perimeter <- function(shape) UseMethod("perimeter")
perimeter.Shape <- function(shape) shape$perimeter

# ============================================================
# R6 Classes: Reference-based OOP
# ============================================================
library(R6)

R6Shape <- R6Class("R6Shape",
  public = list(
    name = NULL,
    color = NULL,
    x = 0,
    y = 0,

    initialize = function(name, color = "black", x = 0, y = 0) {
      self$name <- name
      self$color <- color
      self$x <- x
      self$y <- y
    },

    area = function() stop("Not implemented"),
    perimeter = function() stop("Not implemented"),

    describe = function() {
      paste0(self$name, " [", self$color, "] at (", self$x, ",", self$y, "): ",
             "area=", round(self$area(), 2),
             ", perimeter=", round(self$perimeter(), 2))
    },

    print = function(...) {
      cat(self$describe(), "\n")
      invisible(self)
    }
  )
)

R6Circle <- R6Class("R6Circle",
  inherit = R6Shape,
  public = list(
    radius = NULL,

    initialize = function(radius, color = "black", x = 0, y = 0) {
      super$initialize("Circle", color, x, y)
      self$radius <- radius
    },

    area = function() pi * self$radius^2,
    perimeter = function() 2 * pi * self$radius
  )
)

R6Rectangle <- R6Class("R6Rectangle",
  inherit = R6Shape,
  public = list(
    width = NULL,
    height = NULL,

    initialize = function(width, height, color = "black", x = 0, y = 0) {
      super$initialize("Rectangle", color, x, y)
      self$width <- width
      self$height <- height
    },

    area = function() self$width * self$height,
    perimeter = function() 2 * (self$width + self$height)
  )
)

R6Triangle <- R6Class("R6Triangle",
  inherit = R6Shape,
  public = list(
    a = NULL, b = NULL, c = NULL,

    initialize = function(a, b, c, color = "black", x = 0, y = 0) {
      super$initialize("Triangle", color, x, y)
      self$a <- a
      self$b <- b
      self$c <- c
    },

    area = function() {
      s <- (self$a + self$b + self$c) / 2
      sqrt(s * (s - self$a) * (s - self$b) * (s - self$c))
    },
    perimeter = function() self$a + self$b + self$c
  )
)

# ============================================================
# Demo: S3 classes
# ============================================================
cat("=== S3 Shape Hierarchy ===\n\n")

shapes_s3 <- list(
  Circle(5, "red"),
  Rectangle(4, 6, "blue"),
  Triangle(3, 4, 5, "orange")
)

cat("All Shapes:\n")
cat(strrep("-", 60), "\n")
for (s in shapes_s3) {
  print(s)
}

cat("\nSorted by area (ascending):\n")
areas <- sapply(shapes_s3, function(s) s$area)
sorted <- shapes_s3[order(areas)]
for (s in sorted) {
  cat("  ", s$name, ":", round(s$area, 2), "\n")
}

cat("\nTotal area:", round(sum(areas), 2), "\n")

# ============================================================
# Demo: R6 classes
# ============================================================
cat("\n=== R6 Shape Hierarchy ===\n\n")

shapes_r6 <- list(
  R6Circle$new(5, "red"),
  R6Rectangle$new(4, 6, "blue"),
  R6Triangle$new(3, 4, 5, "orange")
)

cat("All Shapes:\n")
cat(strrep("-", 60), "\n")
for (s in shapes_r6) {
  s$print()
}

cat("\nSorted by area (ascending):\n")
r6_areas <- sapply(shapes_r6, function(s) s$area())
sorted_r6 <- shapes_r6[order(r6_areas)]
for (s in sorted_r6) {
  cat("  ", s$name, ":", round(s$area(), 2), "\n")
}

cat("\nTotal area:", round(sum(r6_areas), 2), "\n")

File: 01-Programming-Languages/Ruby/OOP/classes.rb
# Shape Hierarchy in Ruby
#
# Demonstrates Ruby OOP: classes, modules (mixins),
# inheritance, open classes, and metaprogramming.
#
# Run: ruby classes.rb

# ============================================================
# Modules: Reusable behavior (mixins)
# ============================================================
module Drawable
  def draw
    "Drawing a #{type.downcase} with area #{area.round(2)}"
  end
end

module Calculable
  def self.included(base)
    base.extend(ClassMethods)
  end

  module ClassMethods
    def from_dimensions(*dimensions)
      new(*dimensions)
    end
  end

  def summary
    "#{type}: area=#{area.round(2)}, perimeter=#{perimeter.round(2)}"
  end
end

module Measurable
  def larger_than?(other)
    area > other.area
  end

  def smaller_than?(other)
    area < other.area
  end
end

# ============================================================
# Abstract class (Ruby doesn't enforce, but convention is clear)
# ============================================================
class Shape
  include Drawable
  include Calculable
  include Measurable

  attr_reader :x, :y, :color

  def initialize(x: 0, y: 0, color: "black")
    @x = x
    @y = y
    @color = color
  end

  def area
    raise NotImplementedError, "#{self.class} must implement #area"
  end

  def perimeter
    raise NotImplementedError, "#{self.class} must implement #perimeter"
  end

  def type
    self.class.name
  end

  def to_s
    "#{type} at (#{@x}, #{@y}) [#{color}]: area=#{area.round(2)}, perimeter=#{perimeter.round(2)}"
  end

  def <=>(other)
    area <=> other.area
  end
end

# ============================================================
# Concrete shapes
# ============================================================
class Circle < Shape
  attr_reader :radius

  def initialize(radius, x: 0, y: 0, color: "black")
    super(x: x, y: y, color: color)
    @radius = radius
  end

  def area
    Math::PI * @radius ** 2
  end

  def perimeter
    2 * Math::PI * @radius
  end

  def draw
    "Drawing circle r=#{@radius} at (#{@x}, #{@y})"
  end
end

class Rectangle < Shape
  attr_reader :width, :height

  def initialize(width, height, x: 0, y: 0, color: "black")
    super(x: x, y: y, color: color)
    @width = width
    @height = height
  end

  def area
    @width * @height
  end

  def perimeter
    2 * (@width + @height)
  end
end

class Square < Rectangle
  def initialize(side, x: 0, y: 0, color: "black")
    super(side, side, x: x, y: y, color: color)
  end

  def type
    "Square"
  end

  def side
    @width
  end
end

class Triangle < Shape
  attr_reader :a, :b, :c

  def initialize(a, b, c, x: 0, y: 0, color: "black")
    super(x: x, y: y, color: color)
    @a = a
    @b = b
    @c = c
  end

  def area
    s = (@a + @b + @c) / 2.0
    Math.sqrt(s * (s - @a) * (s - @b) * (s - @c))
  end

  def perimeter
    @a + @b + @c
  end
end

# ============================================================
# Shape collection using Ruby's Enumerable
# ============================================================
class ShapeCollection
  include Enumerable

  def initialize
    @shapes = []
  end

  def add(shape)
    @shapes << shape
    self
  end

  def each(&block)
    @shapes.each(&block)
  end

  def sorted_by_area
    @shapes.sort_by(&:area)
  end

  def total_area
    @shapes.sum(&:area)
  end
end

# ============================================================
# Metaprogramming: Dynamic shape factory
# ============================================================
class ShapeFactory
  def self.create(type, *args)
    shape_class = Object.const_get(type)
    shape_class.new(*args)
  end

  # Dynamic method for each shape type
  %w[Circle Rectangle Square Triangle].each do |shape_name|
    define_singleton_method("create_#{shape_name.downcase}") do |*args|
      create(shape_name, *args)
    end
  end
end

# ============================================================
# Demo
# ============================================================
puts "=== Ruby Shape Hierarchy ==="
puts

# Create shapes
shapes = [
  Circle.new(5, x: 0, y: 0, color: "red"),
  Rectangle.new(4, 6, x: 1, y: 1, color: "blue"),
  Square.new(3, x: 2, y: 2, color: "green"),
  Triangle.new(3, 4, 5, x: 3, y: 3, color: "orange"),
]

# Display all shapes
puts "All Shapes:"
puts "-" * 60
shapes.each { |s| puts s }

# Demonstrate modules
puts "\nDrawing shapes:"
shapes.each { |s| puts "  #{s.draw}" }

# Demonstrate Measurable module
puts "\nComparisons:"
circle = shapes[0]
rectangle = shapes[1]
puts "  Circle larger than Rectangle? #{circle.larger_than?(rectangle)}"
puts "  Rectangle smaller than Triangle? #{rectangle.smaller_than?(shapes[3])}"

# Demonstrate sorting
puts "\nSorted by area (ascending):"
shapes.sort_by(&:area).each { |s| puts "  #{s.type}: #{s.area.round(2)}" }

# Demonstrate collection
puts "\nShape collection:"
collection = ShapeCollection.new
shapes.each { |s| collection.add(s) }
puts "  Total shapes: #{collection.count}"
puts "  Total area: #{collection.total_area.round(2)}"

# Demonstrate metaprogramming factory
puts "\nMetaprogramming factory:"
factory_circle = ShapeFactory.create_circle(3)
puts "  Factory circle: #{factory_circle}"
factory_square = ShapeFactory.create_square(4)
puts "  Factory square: #{factory_square}"

# Demonstrate polymorphism
puts "\nPolymorphism demo:"
def print_shape_info(shape)
  puts "  #{shape.summary}"
end

shapes.each { |s| print_shape_info(s) }
